File: lib/biotp-swift/Sources/BIOTP/OTPGenerator.swift
import Foundation
import CryptoKit

public struct OTPGenerator: Sendable {
    public static let period: TimeInterval = 30
    public static let digits = 6

    public static func generate(sharedSecret: Data, time: Date = Date()) -> String {
        let counter = Self.counter(for: time)
        return generate(sharedSecret: sharedSecret, counter: counter)
    }

    public static func generate(sharedSecret: Data, counter: UInt64) -> String {
        var bigEndian = counter.bigEndian
        let counterData = Data(bytes: &bigEndian, count: 8)

        let key = SymmetricKey(data: sharedSecret)
        let hmac = HMAC<SHA256>.authenticationCode(for: counterData, using: key)
        let hmacBytes = Array(hmac)

        // RFC 4226 dynamic truncation
        let offset = Int(hmacBytes[hmacBytes.count - 1] & 0x0f)
        let truncated = (UInt32(hmacBytes[offset]) & 0x7f) << 24
            | UInt32(hmacBytes[offset + 1]) << 16
            | UInt32(hmacBytes[offset + 2]) << 8
            | UInt32(hmacBytes[offset + 3])

        let modulus = UInt32(pow(10.0, Double(digits)))
        let otp = truncated % modulus
        return String(format: "%0\(digits)d", otp)
    }

    public static func counter(for time: Date = Date()) -> UInt64 {
        UInt64(time.timeIntervalSince1970 / period)
    }

    public static func secondsRemaining(for time: Date = Date()) -> Int {
        Int(period) - Int(time.timeIntervalSince1970.truncatingRemainder(dividingBy: period))
    }
}

File: lib/biotp-kotlin/src/main/kotlin/com/biotp/OTPGenerator.kt
package com.biotp

import java.nio.ByteBuffer
import java.security.MessageDigest
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlin.math.pow

/** HOTP/TOTP generation and X9.63 KDF for the BIOTP protocol. */
object OTPGenerator {
    const val PERIOD: Long = 30
    const val DIGITS: Int = 6

    fun generate(sharedSecret: ByteArray, counter: Long): String {
        val counterBytes = ByteBuffer.allocate(8).putLong(counter).array()
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(sharedSecret, "HmacSHA256"))
        val hs = mac.doFinal(counterBytes)

        // RFC 4226 dynamic truncation
        val offset = (hs[hs.size - 1].toInt() and 0x0F)
        val binary = ((hs[offset].toInt() and 0x7F) shl 24) or
            ((hs[offset + 1].toInt() and 0xFF) shl 16) or
            ((hs[offset + 2].toInt() and 0xFF) shl 8) or
            (hs[offset + 3].toInt() and 0xFF)

        val modulus = 10.0.pow(DIGITS).toInt()
        val otp = binary % modulus
        return otp.toString().padStart(DIGITS, '0')
    }

    fun counter(timeSeconds: Long = System.currentTimeMillis() / 1000): Long {
        return timeSeconds / PERIOD
    }

    fun secondsRemaining(timeSeconds: Long = System.currentTimeMillis() / 1000): Int {
        return (PERIOD - (timeSeconds % PERIOD)).toInt()
    }

    /**
     * X9.63 KDF with SHA-256, empty SharedInfo, 32-byte output.
     *
     * `derived = SHA-256(Z || 00000001)`
     */
    fun x963kdf(sharedSecret: ByteArray): ByteArray {
        val md = MessageDigest.getInstance("SHA-256")
        md.update(sharedSecret)
        md.update(byteArrayOf(0, 0, 0, 1))
        return md.digest()
    }
}
